File: 020_Valid_parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        for (int i = 0; i < s.length(); i++){
            if (isLeftParenthesis(s.charAt(i))) {
                stack.push(s.charAt(i));
            }else{
                Character candidate = null;
                try{
                    candidate = stack.pop();
                } catch (Exception e){
                    return false;
                }
                if (isParenthesesMatched(candidate, s.charAt(i))) continue;
                else return false;
            }
        }
        if (! stack.empty()) return false;
        return true;
    }

    private boolean isLeftParenthesis(char c){
        return (c == '{' || c == '[' || c == '(') ? true : false;
    }

    private boolean isParenthesesMatched(char candidate, char matcher){
        if ((candidate == '{' && matcher == '}') ||
            (candidate == '[' && matcher == ']') ||
            (candidate == '(' && matcher == ')')) {
            return true;
        }
        return false;
    }
}

File: python/020_valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stack.append(c)
            else:
                if len(stack) <= 0:
                    return False
                else:
                    target = stack.pop()
                    if not self.checkMatch(target, c):
                        return False
        return True if len(stack) == 0 else False


    def checkMatch(self, target: str, cand: str):
        if target == "(" and cand == ")":
            return True
        elif target == "{" and cand == "}":
            return True
        elif target == "[" and cand == "]":
            return True
        return False
